src/kv/mod.rs

```diff
@@ -0,0 +1,410 @@
+use std::collections::HashMap;
+use std::io::Write;
+
+use async_nats::jetstream::kv::Store;
+use async_nats::Client;
+use data_encoding::HEXUPPER;
+use ring::digest::{digest, SHA256};
+
+use crate::LinkDefinition;
+use crate::Result;
+
+mod cached;
+mod direct;
+
+pub use cached::*;
+pub use direct::*;
+
+const LINKDEF_PREFIX: &str = "LINKDEF_";
+const CLAIMS_PREFIX: &str = "CLAIMS_";
+const BUCKET_PREFIX: &str = "LATTICEDATA_";
+const SUBJECT_KEY: &str = "sub";
+
+#[async_trait::async_trait]
+pub trait KvStore {
+    /// Returns all links in the store
+    async fn get_links(&self) -> Result<Vec<LinkDefinition>>;
+
+    /// Returns all claims in the store
+    async fn get_claims(&self) -> Result<Vec<HashMap<String, String>>>;
+
+    /// Returns all provider claims in the store
+    async fn get_provider_claims(&self) -> Result<Vec<HashMap<String, String>>>;
+
+    /// Returns all actor claims in the store
+    async fn get_actor_claims(&self) -> Result<Vec<HashMap<String, String>>>;
+
+    /// Returns all links in the store that match the provided filter function. For some
+    /// implementations, this could be more efficient than fetching all links and filtering them in
+    /// memory.
+    async fn get_filtered_links<F>(&self, filter_fn: F) -> Result<Vec<LinkDefinition>>
+    where
+        F: FnMut(&LinkDefinition) -> bool + Send;
+
+    /// Returns the link definition for the given actor, link name, and contract ID, if it exists.
+    async fn get_link(
+        &self,
+        actor_id: &str,
+        link_name: &str,
+        contract_id: &str,
+    ) -> Result<Option<LinkDefinition>>;
+
+    /// Returns the claim for the given ID, if it exists.
+    async fn get_claim(&self, id: &str) -> Result<Option<HashMap<String, String>>>;
+
+    /// Adds a link definition to the store
+    async fn put_link(&self, ld: LinkDefinition) -> Result<()>;
+
+    /// Deletes a link definition from the store
+    async fn delete_link(&self, actor_id: &str, contract_id: &str, link_name: &str) -> Result<()>;
+
+    // TODO(thomastaylor312): Once we're ready, we should probably add the put_claim and
+    // delete_claim logic here from the host
```

Comment:
What are the requirements to be ready for this?

Reply:
I consider this a lower priority, quality of life enhancement to the host (since only the host puts claims). Now that I am typing that, maybe we shouldn't ever have it in the control client since it isn't something that clients do. Thoughts?

Reply:
Discussed offline: writing/deleting claims should only be done by the host, so we can leave this out